File: Backtracking/nqueen.c
/* Another N-Queen Problem*/
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

long long lineCounter;

int place(int queen, int col, int dim, int *x, int bad[]){

	int i, j;
	//
	//printf("Bad Squares:");
	//for(i=0; i < dim; i++){
	//	printf(" %d", board[i]);
	//}printf("\n");
	
	int rowBit = 1 << dim - col;
	int badBit = bad[queen-1];
	int result = rowBit & badBit;
	if (result != 0){ /* This is on a *BAD SQUARE!*  */
		return 0;
	}

	
	/* Verify badx, bady*//*
	printf("Bad squares \n");
	for (i = 0; i < bad_len; i++){
		printf(" (%d, %d)", badx[i], bady[i]);
	}printf("\n");*/
	

	int prev;
	for (prev = 1; prev < queen; prev++){
		if ((x[prev] == col) || ((abs(x[prev]-col) == abs(prev-queen)))){			
			if (x[prev] == col){
				//printf("x:");
				//for (i=0; i< queen+1; i++){
				//	printf(" %d", x[i]);
				//}printf("\n"); 	

				//printf("Placing queen %d on col %d...", queen, col);
				//printf("Failed! x[%d]:%d == %d\n\n", prev, x[prev], col);
			}
			//else{
				//printf("failed on diagonal\n");
			//}
			return 0;
		}
	}
	//printf("Succeeded!\n");
	return 1;	
}

int NQueens(int queen, int dim, int *x, int bad[]){
	int i, j, row;
	for (row = 1; row < dim+1; row++){
		if((place(queen, row, dim, x, bad)== 1)){
			x[queen] = row;
			if (queen == dim){				
				/*printf("x:");
				for (i=0; i< queen+1; i++){
					printf(" %d", x[i]);
				}printf("\n");*/
				lineCounter = lineCounter + 1;
			}
			else{				
				NQueens(queen + 1, dim, x, bad); /*recursively try next position*/
			}
		}
	}
	return (lineCounter);
}


int main(int argc, char** argv){
	int rc;
	int dim;
	int case_num = 1;
	rc = scanf("%d", &dim);
	while (dim != 0){
		int i, j, k;
		int bad[dim];
		for (i = 0; i < dim; i++){
			bad[i] = 0;
		}
		/* Zero out x!!*/
		int x[dim+1];
		for (j = 0; j < dim+1; j++){
			x[j] = -1;
		}
		char boardline[dim+1];
		int boardMax = 1 << dim-1;
		//printf("Board Max: %d\n", boardMax);
		int val;
		/* Find the bad squares on the board*/
		for (k = 0; k < dim; k++){
			rc = scanf("%s", boardline);
			char chr;

			for (j = 0; j < dim; j++){
				chr = boardline[j];
				if (chr == '*'){
					val = boardMax >> j;
					bad[k] = bad[k] + val;
				}
			}
		}
		/* Verify board:*//*
		printf("Bad Squares:");
		for(i=0; i < dim; i++){
			printf(" %d", board[i]);
		}printf("\n");*/

		
		lineCounter = 0;
		if (dim == 3){ /* No configurations on a 3x3 board*/
			printf("Case %d: %d\n",case_num, 0);
		}
		else{
			int total;
			total = NQueens(1, dim, x, bad);
			printf("Case %d: %d\n", case_num, total);
		}
		rc = scanf("%d", &dim);
		case_num = case_num + 1;
	}
	return 0;
}

File: Backtracking/n_queen.py
# Another N-Queen Problem

import sys

sys.stdin = open("input.txt")
import sys

lineCounter = None
x = [None]

def place(queen, row, bad):
	global x
	#print(x)
	for prev in range(1, queen): #check previously placed queens
		if (x[prev] == row or (abs(x[prev]-row) == abs(prev-queen))):
			return False
	for i, j in bad:
		if queen == i and row == j:
			return False
	return True

def NQueens(queen, dim, bad_squares):
	global x
	global lineCounter

	for row in range(1, dim+1):
		if(place(queen, row, bad_squares)):
			x[queen] = row
			if (queen == dim):
				#msg = str(lineCounter) + "      " + str(x[1])
				#if (lineCounter < 10):
				#	msg = " " + msg
				#sys.stdout.write(msg)
				
				#for j in range(2, dim+1):
				#	sys.stdout.write(" " +str(x[j]))
				lineCounter = lineCounter + 1

				#sys.stdout.write("\n")
			else:
				NQueens(queen + 1, dim, bad_squares) # recursively try next position

	return (lineCounter-1)

def load():
	dim = int(next(sys.stdin))
	while (dim != 0):
		board = list()
		for i in range(0, dim):
			line = next(sys.stdin)
			j = 0
			for char in line:
				if char == '*':
					board.append((i+1, j+1))
				j = j + 1
		yield(dim, board)
		dim = int(next(sys.stdin))

def main():
	global lineCounter
	global x 
	i = 1
	for dim, board in load():
		x = [None]* (dim+1)
		#print("Board: "+ str(board))
		placement = [i for i in range(0, dim)]
		lineCounter = 1
		#sys.stdout.write("SOLN       COLUMN\n")
		#sys.stdout.write(" #      1 2 3 4 5 6 7 8\n\n")
		total = NQueens(1, dim, board)
		sys.stdout.write("Case {}: {}\n".format(i, total))
		i = i + 1
		#make nqueens accept generic input
		#make nqueens care about board size
		#make nqueens care about bad spaces

main()